packages/frontend/core/src/components/filter/filters.tsx

```diff
@@ -9,11 +10,21 @@ export const Filters = ({
   filters,
   className,
   onChange,
+  defaultDraftFilter,
 }: {
   filters: FilterParams[];
   className?: string;
   onChange?: (filters: FilterParams[]) => void;
+  defaultDraftFilter?: FilterParams | null;
 }) => {
+  const [draftFilter, setDraftFilter] = useState<FilterParams | null>(
+    defaultDraftFilter ?? null
+  );
+  const draftFilterRef = useRef<FilterParams | null>(draftFilter);
+  useEffect(() => {
+    draftFilterRef.current = draftFilter;
```

Comment:
what's this for?

Reply:
When draftChange and draftCompleted are triggered consecutively,
we might save an outdated draft filter value.
Using a ref helps us avoid this issue by always accessing the latest value.

Reply:
you can directly use `draftFilter` in hooks and without putting it in deps array. it's closure.

Reply:
the useEffect is obviously redundant

Reply:
Because it is a closure, the `handleDraftCompleted` callback will remember the `draftFilter` value at render time.

Reply:
```
  React.useEffect(() => {
    callbackRef.current = callback;
  });
```
this is pure shit

Reply:
their react code written like shit

Reply:
The render function should be pure, so we can only write side effects in useEffect